src/clj_gcloud/storage.clj

```diff
@@ -105,9 +105,22 @@
     (.build builder)))
 
 (defn copy
-  [^Storage storage ^BlobId source-blob-id ^BlobId target-blob-id]
-  (-> (.copy storage (Storage$CopyRequest/of source-blob-id target-blob-id))
-      (.getResult)))
+  ([storage source-blob-id target-blob-id]
+   (copy storage source-blob-id target-blob-id nil))
+  ([^Storage storage ^BlobId source-blob-id ^BlobId target-blob-id options]
+   (let [cr (if (:with-precondition options)
+              ;; Copy blobs within cloud storage using preconditions recommended in Google Cloud Storage guides
+              ;; See: https://cloud.google.com/storage/docs/copying-renaming-moving-objects#storage-copy-object-java
+              (let [precondition (if-let [target (.get storage target-blob-id)]
+                                   (Storage$BlobTargetOption/generationMatch (.getGeneration target))
+                                   (Storage$BlobTargetOption/doesNotExist))]
+                (-> (Storage$CopyRequest/newBuilder)
+                    (.setSource source-blob-id)
+                    (.setTarget target-blob-id (into-array Storage$BlobTargetOption [precondition]))
```

Comment:
To avoid a reflection warning on this line because `into-array` does not work properly. We could use the `iterable` arity of `.setTarget`. In order to do that we can add in the `let` a new binding to a hinted symbol (see below) and use it in  `.setTarget`.

```clojure
^Iterable target-opts [precondition]
```